util/linuxfw/iptables.go

```diff
@@ -823,3 +825,479 @@ type markTarget struct {
+}
+
+func NewIptableRunner(logf logger.Logf) (*iptableRunner, error) {
+	ipt4, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		return nil, err
+	}
+
+	v6err := checkIPv6(logf)
+	if v6err != nil {
+		logf("disabling tunneled IPv6 due to system IPv6 config: %v", v6err)
+	}
+	supportsV6 := v6err == nil
+	supportsV6NAT := supportsV6 && supportsV6NAT()
+	if supportsV6 {
+		logf("v6nat = %v", supportsV6NAT)
+	}
+
+	var ipt6 *iptables.IPTables
+	if supportsV6 {
+		ipt6, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newIptableRunnerAdvanced(ipt4, ipt6, supportsV6, supportsV6NAT), nil
+}
+
+// newIptableRunnerAdvanced is a helper function for testing purposes
+func newIptableRunnerAdvanced(ipt4, ipt6 iptablesInterface, supV6, supV6NAT bool) *iptableRunner {
+	return &iptableRunner{ipt4, ipt6, supV6, supV6NAT}
+}
+
+func (i *iptableRunner) GetV6Available() bool {
+	return i.v6Available
+}
+
+func (i *iptableRunner) GetV6NATAvailable() bool {
+	return i.v6NATAvailable
+}
+
+// AddLoopbackRule adds a iptables rule to permit loopback traffic to
+// a local Tailscale IP.
+func (i *iptableRunner) AddLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Insert("filter", "ts-input", 1, "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("adding loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// tsChain returns the name of the tailscale sub-chain corresponding
+// to the given "parent" chain (e.g. INPUT, FORWARD, ...).
+func tsChain(chain string) string {
+	return "ts-" + strings.ToLower(chain)
+}
+
+// getIptablebyProto returns the iptablesInterface in iptables runner for the given protocol.
+func (i *iptableRunner) getIptablebyProto(proto ProtoVersion) iptablesInterface {
+	switch proto {
+	case IPv4:
+		return i.ipt4
+	case IPv6:
+		return i.ipt6
+	default:
+		panic(fmt.Sprintf("unknown protocol %v", proto))
+	}
+}
+
+// DelLoopbackRule removes the iptables rule permitting loopback
+// traffic to a Tailscale IP.
+func (i *iptableRunner) DelLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Delete("filter", "ts-input", "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("deleting loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// GetTables gets the available iptablesInterface in iptables runner.
+func (i *iptableRunner) getTables() []iptablesInterface {
+	if i.GetV6Available() {
+		return []iptablesInterface{i.ipt4, i.ipt6}
+	}
+	return []iptablesInterface{i.ipt4}
+}
+
+// addNetfilterHooks inserts calls to tailscale's netfilter chains in
+// the relevant main netfilter chains. The tailscale chains must
+// already exist.
+func (i *iptableRunner) AddNetfilterHooks() error {
+	divert := func(ipt iptablesInterface, table, chain string) error {
```

Comment:
Whats the function of `divert()`? What is it diverting?

Reply:
It was named divert before I changed it so just kept the name. Is there a better name you would suggest?